File: src/main/java/uk/co/optimisticpanda/jnomon/formatter/AbsoluteOutputWriter.java
package uk.co.optimisticpanda.jnomon.formatter;

import static uk.co.optimisticpanda.jnomon.Utils.MOVE_TO_START_OF_PREV_LINE;
import static uk.co.optimisticpanda.jnomon.Utils.formatLine;
import static uk.co.optimisticpanda.jnomon.Utils.formatSeconds;
import static uk.co.optimisticpanda.jnomon.Utils.nowAsTimestamp;

import java.time.Duration;

import uk.co.optimisticpanda.jnomon.Utils.Colour;

public class AbsoluteOutputWriter implements EventListener {

    @Override
    public void onLineStart(Colour colour, Duration sinceProcessStart, Duration sinceLastStart, String line) {
        onUpdate(colour, sinceProcessStart, sinceLastStart);
        printer().println(formatLine(24, "", line));
    }
    
    @Override
    public void onUpdate(Colour colour, Duration sinceProcessStart, Duration sinceLastStart) {
        String text = formatLine(24, nowAsTimestamp(), "");
        printer().println(MOVE_TO_START_OF_PREV_LINE + colour.colourize(text));
    }
    
    @Override
    public void onLineEnd(Duration sinceProcessStart, Duration sinceLastStart, String lastLine) {
        printer().println(formatLine(24, nowAsTimestamp(), lastLine));
    }
    
    @Override
    public void onFinally(Duration sinceProcessStart, Duration sinceLastStart) {
        printer().println(formatLine(24, "Total:", formatSeconds(sinceProcessStart)));
    }
}

File: src/main/java/uk/co/optimisticpanda/jnomon/formatter/AbsoluteOutputWriter.kt
package uk.co.optimisticpanda.jnomon.formatter

import uk.co.optimisticpanda.jnomon.Utils.Colour
import uk.co.optimisticpanda.jnomon.Utils.MOVE_TO_START_OF_PREV_LINE
import uk.co.optimisticpanda.jnomon.Utils.formatLine
import uk.co.optimisticpanda.jnomon.Utils.formatSeconds
import uk.co.optimisticpanda.jnomon.Utils.nowAsTimestamp
import java.time.Duration

class AbsoluteOutputWriter : EventListener {
  override fun onLineStart(colour: Colour?, sinceProcessStart: Duration, sinceLastStart: Duration, line: String?) {
    onUpdate(colour, sinceProcessStart, sinceLastStart)
    printer().println(formatLine(24, "", line))
  }

  override fun onUpdate(colour: Colour?, sinceProcessStart: Duration, sinceLastStart: Duration) {
    val text: String = formatLine(24, nowAsTimestamp(), "")
    printer().println(MOVE_TO_START_OF_PREV_LINE + colour!!.colourize(text))
  }

  override fun onLineEnd(sinceProcessStart: Duration, sinceLastStart: Duration, lastLine: String?) {
    printer().println(formatLine(24, nowAsTimestamp(), lastLine))
  }

  override fun onFinally(sinceProcessStart: Duration, sinceLastStart: Duration?) {
    printer().println(formatLine(24, "Total:", formatSeconds(sinceProcessStart)))
  }
}